### PR TITLE
Web console: editing compute resource limits

### PR DESCRIPTION
**Pull request by spadgett**
* Allow editing compute resource requests and limits for deployment configs and replication controllers that don't have a deployment config.
* Allow setting compute resource requests and limits during create.

https://trello.com/c/23NdtCwN

- [x] Add compute resources to create form
- [x] Disable form on submit
- [x] Update application generator unit tests
- [x] Update for #6901
- [x] Don't require limit if a default is set
- [x] https://bugzilla.redhat.com/show_bug.cgi?id=1303547

Browse deployment config:

openshift_web_console

Set resource limits page:

openshift_web_console

/cc @jwforres @derekwaynecarr @sosiouxme

### CODE REVIEW

**Comment by spadgett** [comment]
@jwforres PTAL

File: assets/app/scripts/controllers/create/createFromImage.js [hunk 1]
```diff
@@ -99,6 +102,12 @@ angular.module("openshiftConsole")
             });
         }
 
+        var watches = [];
+        watches.push(DataService.watch("limitranges", context, function(limitRanges) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/create/createFromImage.js]: limit ranges very very rarely change, wondering if we should just list here and save a websocket connection,  when we originally talked about watching for this stuff it was for quota which was going to be updating potentially.

Comment (spadgett) [inline assets/app/scripts/controllers/create/createFromImage.js]: Yeah, I'll make the change.

**Comment by spadgett** [comment]
@jwforres thoughts on this alternative for showing limit range and default?

openshift_web_console

**Comment by spadgett** [comment]
Hm, maybe the placeholder is a bad idea for the default.

File: assets/app/scripts/controllers/setLimits.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                });
+            };
+          },
+          function(result) {
+            displayError(displayName + ' could not be loaded.', getErrorDetails(result));
+          }
+        );
+
+        var watches = [];
+        watches.push(DataService.watch("limitranges", context, function(limitRanges) {
```
Comment (jwforres) [inline assets/app/scripts/controllers/setLimits.js]: same thing here, change to a list

**Comment by spadgett** [comment]
Rebased and removed the limit range watches.

File: assets/app/scripts/directives/editRequestLimit.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+          break;
+        }
+
+        var usageValue = $filter('usageValue');
+        var amountAndUnit = $filter('amountAndUnit');
+
+        var validateLimitRange = function() {
+          var value, minValid = true, maxValid = true;
+          if (scope.amount) {
+            value = usageValue(scope.amount + scope.unit, scope.unit);
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: trying to follow exactly where unit normalization is happening with the current value vs the min and max.  some comments might help here

**Comment by spadgett** [comment]
Rebased and updated. Current UI:

openshift_web_console

@jwforres please have another look

**Comment by spadgett** [comment]
[test]

File: assets/app/scripts/directives/editRequestLimit.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+            label: "MiB"
+          }, {
+            value: "Gi",
+            label: "GiB"
+          }];
+          break;
+        }
+
+        var validateLimitRange = function() {
+          // Use usageValue filter to normalize units for comparison.
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: thanks this section is a lot easier to follow now

File: assets/app/scripts/directives/editRequestLimit.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+              limitLargerThanRequest = true,
+              limitWithinRatio = true;
+
+          if (request && limit) {
+            // Limit must be greater than or equal to request.
+            limitLargerThanRequest = limit >= request;
+
+            // Limit must be within the max limit/request ratio if defined.
+            if (scope.maxLimitRequestRatio) {
+              limitWithinRatio = (limit / request) <= scope.maxLimitRequestRatio;
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: when request is set and limit isnt then you have an unbounded limit, if a a maxLimitRequestRatio is set then this pod would be rejected ( i just double checked with @derekwaynecarr )

File: assets/app/scripts/directives/editRequestLimit.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+          // Normalize units to compare.
+          return usageValue(candidate) > usageValue(previous);
+        };
+
+        var updateLimitRanges = function() {
+          scope.limits = {};
+          angular.forEach(scope.limitRanges, function(limitRange) {
+            angular.forEach(limitRange.spec.limits, function(limit) {
+              // Find the largest minimum.
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: should this be in a LimitRanges service, LimitRanges.getEffectiveLimits() or something.  Feels like we are going to need this somewhere else.

File: assets/app/scripts/directives/editRequestLimit.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+              if (limit.min && isLarger(limit.min[scope.type], scope.cpuMin)) {
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: scope.min instead of scope.cpuMin?  since it might be the memory min right?

Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: yeah this seems to be a typo, you are setting scope.min below

Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: Yeah, it's a typo. I'll fix.

File: assets/app/scripts/directives/editRequestLimit.js [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                scope.min = limit.min[scope.type];
+              }
+              if (limit.max && isSmaller(limit.max[scope.type], scope.max)) {
+                scope.max = limit.max[scope.type];
+              }
+              if (limit['default']) {
+                scope.defaultLimit = limit['default'][scope.type] || scope.defaultLimit;
```
Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: add a TODO here that we need to revisit this when we decide what the correct indeterminate behavior is for multiple valid defaults

File: assets/app/scripts/filters/util.js [hunk 9]
```diff
@@ -105,6 +105,12 @@ angular.module('openshiftConsole')
         // We didn't get an amount? shouldn't happen but just in case
         return [value, null];
       }
+
+      // No normalization of unit.
+      if (!humanizeUnits) {
+        return split.slice(1);
```
Comment (jwforres) [inline assets/app/scripts/filters/util.js]: So this is stripping the unit and only returning the value?  I think I would rather have usageValue take a flag that prevents normalization.  It's weird that the default behavior of a function called amountAndUnit is to drop the unit.

File: assets/app/styles/_core.less [hunk 10]
```diff
@@ -517,6 +517,15 @@ label.checkbox {
   }
 }
 
+.compute-resource .inline-select {
+  // Spacing below input at mobile sizes.
+  margin-top: 5px;
```
Comment (jwforres) [inline assets/app/styles/_core.less]: Looks like this is always added, not just at mobile, fix comment?

Comment (spadgett) [inline assets/app/styles/_core.less]: No, it's reset to 0 just below at screen-sm-min.

File: assets/app/scripts/filters/util.js (hunk 9 above)
Comment (spadgett) [inline assets/app/scripts/filters/util.js]: > So this is stripping the unit and only returning the value?

No, it's returning the amount (index 1) and unit (index 2).

openshift_web_console

Comment (spadgett) [inline assets/app/scripts/filters/util.js]: I'll add a comment.

Comment (jwforres) [inline assets/app/scripts/filters/util.js]: ah duh ok

Comment (spadgett) [inline assets/app/scripts/filters/util.js]: It's confusing. I had to set a breakpoint to remember why I needed the slice.

Comment (spadgett) [inline assets/app/scripts/filters/util.js]: I've updated the code so things are more clear.

File: assets/app/scripts/directives/editRequestLimit.js (hunk 8 above)
Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: Added the TODO comment.

File: assets/app/scripts/directives/editRequestLimit.js (hunk 6 above)
Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: Added a service.

File: assets/app/scripts/directives/editRequestLimit.js (hunk 5 above)
Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: Fixed.

**Comment by spadgett** [comment]
@jwforres comments addressed

**Comment by jwforres** [comment]
LGTM fork_ami

**Comment by spadgett** [comment]
Rebased and fixed java/bindata conflict.

**Comment by spadgett** [comment]
Rebased.

**Comment by sosiouxme** [comment]
@spadgett FYI I think the PodLimitRequest mechanics are firmed up, will find out more solidly at scrum this afternoon.

See implementation details under https://trello.com/c/rRN7m0zv

PR https://github.com/openshift/origin/pull/6901

File: assets/app/scripts/directives/editRequestLimit.js (hunk 5 above)
Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: @jwforres It looks like limit is not required when there's a default, even with `maxLimitRequestRatio` set.

Comment (jwforres) [inline assets/app/scripts/directives/editRequestLimit.js]: Yeah it is only an issue when there is no default

Comment (spadgett) [inline assets/app/scripts/directives/editRequestLimit.js]: I'll fix.

**Comment by spadgett** [comment]
Rebased.

**Comment by spadgett** [comment]
[test]

File: assets/app/scripts/services/limits.js [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+          min = podLimits.min && usageValue(podLimits.min),
+          max = podLimits.max && usageValue(podLimits.max),
+          problems = [],
+          computeResourceLabel = $filter('computeResourceLabel')(computeResource, true);
+
+      // Total the usage for all containers.
+      angular.forEach(containers, function(container) {
+        var resources = container.resources;
+        if (!resources) {
+          return;
```
Comment (jwforres) [inline assets/app/scripts/services/limits.js]: wouldn't this container's defaults still matter, you wouldn't have validation on these fields but the defaults should still go against the total

**Comment by jwforres** [comment]
LGTM [merge]

File: assets/app/scripts/controllers/setLimits.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+      displayName = 'Deployment Configuration "' + $scope.name + '"';
+      $scope.resourceURL = Navigate.resourceURL($scope.name, "DeploymentConfig", $routeParams.project);
+    } else if ($routeParams.rcName) {
+      type = "replicationcontrollers";
+      $scope.name = $routeParams.rcName;
+      displayName = 'Replication Controller "' + $scope.name + '"';
+      $scope.resourceURL = Navigate.resourceURL($scope.name, "ReplicationController", $routeParams.project);
+      $scope.showPodWarning = true;
+    } else {
+      Navigate.toErrorPage("A replication controller or deployment config must be provided.");
```
Comment (liggitt) [inline assets/app/scripts/controllers/setLimits.js]: return here?

File: assets/app/scripts/controllers/setLimits.js [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+ */
+angular.module('openshiftConsole')
+  .controller('SetLimitsController', function ($filter, $location, $parse, $routeParams, $scope, AlertMessageService, DataService, LimitRangesService, Navigate, ProjectsService) {
+    if ($routeParams.dcName && $routeParams.rcName) {
+      Navigate.toErrorPage("Replication controller and deployment config can't both be provided.");
+    }
+
+    $scope.hideCPU = LimitRangesService.hideCPUComputeResources();
+
+    var type, displayName;
```
Comment (liggitt) [inline assets/app/scripts/controllers/setLimits.js]: nit, resource, not type

**Pull request merged**